From: Eric Dumazet <edumazet@google.com>
Date: Fri, 17 Oct 2014 12:45:55 -0700
Subject: [PATCH 46/46] bna: fix skb->truesize underestimation
Patch-mainline: v3.18-rc1
Git-commit: f2d9da1a8375cbe53df5b415d059429013a3a79f
References: bsc#909364 FATE#317497

skb->truesize is not meant to be tracking amount of used bytes
in an skb, but amount of reserved/consumed bytes in memory.

For instance, if we use a single byte in last page fragment,
we have to account the full size of the fragment.

skb->truesize can be very different from skb->len, that has
a very specific safety purpose.

Signed-off-by: Eric Dumazet <edumazet@google.com>
Cc: Rasesh Mody <rasesh.mody@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Gary Ching-Pang Lin <glin@suse.com>
---
 drivers/net/ethernet/brocade/bna/bnad.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/drivers/net/ethernet/brocade/bna/bnad.c
+++ b/drivers/net/ethernet/brocade/bna/bnad.c
@@ -554,6 +554,7 @@ bnad_cq_setup_skb_frags(struct bna_rcb *
 
 		len = (vec == nvecs) ?
 			last_fraglen : unmap->vector.len;
+		skb->truesize += unmap->vector.len;
 		totlen += len;
 
 		skb_fill_page_desc(skb, skb_shinfo(skb)->nr_frags,
@@ -565,7 +566,6 @@ bnad_cq_setup_skb_frags(struct bna_rcb *
 
 	skb->len += totlen;
 	skb->data_len += totlen;
-	skb->truesize += totlen;
 }
 
 static inline void
